pkg/distributor/distributor.go

```diff
@@ -339,6 +398,59 @@ func (d *Distributor) Push(ctx context.Context, req *logproto.PushRequest) (*log
 	}
 }
 
+func min(x1, x2 int) int {
+	if x1 < x2 {
+		return x1
+	}
+	return x2
+}
+
+// shardStream shards (divides) the given stream into smaller streams.
+//
+// It will derive N streams (and its entries) from the given stream, where N is the sharding size for the given stream.
+func shardStream(stream logproto.Stream, cfg Config, streamSharder StreamSharder, userID string) ([]uint32, []streamTracker) {
+	logger := util_log.Logger
+	shards, ok := streamSharder.ShardsFor(stream)
+	if !ok || shards == 1 {
+		return []uint32{util.TokenFor(userID, stream.Labels)}, []streamTracker{{stream: stream}}
+	}
+
+	derivedKeys := make([]uint32, 0, shards)
+	derivedStreams := make([]streamTracker, 0, shards)
+
+	if cfg.ShardStreams.Debug {
+		level.Warn(logger).Log("msg", "sharding request with mode", "mode", cfg.ShardStreams.Mode)
+	}
+
+	entriesPerWindow := float64(len(stream.Entries)) / float64(shards) // divide and keep decimal value.
+	for i := 0; i < shards; i++ {
+		streamCopy := logproto.Stream{}
+		lbs, err := syntax.ParseLabels(stream.Labels)
+		if err != nil {
+			level.Error(logger).Log("msg", "couldn't extract labels from stream", "stream", streamCopy.Labels)
+			continue
+		}
+
+		lbs = append(lbs, labels.Label{Name: ShardLbName, Value: fmt.Sprintf("%d", i)})
+		streamCopy.Labels = lbs.String()
+		streamCopy.Hash = lbs.Hash()
+
+		fIdx := float64(i)
+		lowerBound := int(fIdx * entriesPerWindow)
+		upperBound := min(int(entriesPerWindow*(1+fIdx)), len(stream.Entries))
+
+		streamCopy.Entries = stream.Entries[lowerBound:upperBound]
```

Comment:
I think we should validate here that:

1. `lowerBound >= 0` (will panic otherwise)
2. `lowerBound < upperBound` (otherwise it'll panic if greater, or return no entries if equal)

Reply:
Agreed on the first, pushed a change that guarantees we'll only calculate all of this if shards  > 0, which guards against it.

For second, I wrote a guard-clause for it and added a few tests covering it (number of shards > len(entries) is the only scenario where lowerbound>upperbound would be possible).